aries_cloudagent/admin/decorators/auth.py

```diff
@@ -0,0 +1,80 @@
+"""Authentication decorators for the admin API."""
+
+import functools
+
+from aiohttp import web
+
+from ...utils import general as general_utils
+from ..request_context import AdminRequestContext
+
+
+def admin_authentication(handler):
+    """Decorator to require authentication via admin API key.
+
+    The decorator will check for a valid x-api-key header and
+    reject the request if it is missing or invalid.
+    If the agent is running in insecure mode, the request will be allowed without a key.
+    """
+
+    @functools.wraps(handler)
+    async def wrapper(request):
+        context: AdminRequestContext = request["context"]
+        profile = context.profile
+        header_admin_api_key = request.headers.get("x-api-key")
+        valid_key = general_utils.const_compare(
+            profile.settings.get("admin.admin_api_key"), header_admin_api_key
+        )
+        insecure_mode = bool(profile.settings.get("admin.admin_insecure_mode"))
```

Comment:
Is it better to just shortcut out of the function if `insecure_mode` is set, right at the start of the function?

Reply:
The condition would still need to be checked, so I preferred having only two exit points rather than a bigger if/else block. I can get those two extracted to a first condition if it seems better, but overall they all follow the same logic.